File: sdk/python/src/sigcli_sdk/crypto.py
from __future__ import annotations

import base64
import json
from pathlib import Path
from typing import Any, Optional

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

DEFAULT_SIG_DIR = Path.home() / ".sig"
KEY_FILE = "encryption.key"


def decrypt(envelope: dict[str, Any], key: bytes) -> str:
    iv = base64.b64decode(envelope["iv"])
    auth_tag = base64.b64decode(envelope["authTag"])
    ciphertext = base64.b64decode(envelope["ciphertext"])

    aesgcm = AESGCM(key)
    # AESGCM expects ciphertext + authTag concatenated
    plaintext = aesgcm.decrypt(iv, ciphertext + auth_tag, None)
    return plaintext.decode("utf-8")


def is_encrypted_envelope(data: Any) -> bool:
    if not isinstance(data, dict):
        return False
    return (
        data.get("encrypted") is True
        and data.get("version") == 1
        and data.get("algorithm") == "aes-256-gcm"
        and isinstance(data.get("iv"), str)
        and isinstance(data.get("authTag"), str)
        and isinstance(data.get("ciphertext"), str)
    )


def load_encryption_key(sig_dir: Optional[Path] = None) -> bytes:
    directory = sig_dir or DEFAULT_SIG_DIR
    key_path = directory / KEY_FILE

    if not key_path.exists():
        raise FileNotFoundError(
            f"Encryption key not found at {key_path}. Run 'sig init' to generate one."
        )

    raw = key_path.read_text(encoding="utf-8").strip()
    key = base64.b64decode(raw)
    if len(key) != 32:
        raise ValueError(f"Invalid encryption key: expected 32 bytes, got {len(key)}")

    return key

File: sdk/typescript/src/crypto.ts
import { createDecipheriv } from 'node:crypto';
import fs from 'node:fs/promises';
import path from 'node:path';
import os from 'node:os';

export interface EncryptedEnvelope {
    version: 1;
    encrypted: true;
    algorithm: 'aes-256-gcm';
    iv: string;
    authTag: string;
    ciphertext: string;
}

const DEFAULT_SIG_DIR = path.join(os.homedir(), '.sig');
const KEY_FILE = 'encryption.key';

export function decrypt(envelope: EncryptedEnvelope, key: Buffer): string {
    const iv = Buffer.from(envelope.iv, 'base64');
    const authTag = Buffer.from(envelope.authTag, 'base64');
    const ciphertext = Buffer.from(envelope.ciphertext, 'base64');

    const decipher = createDecipheriv('aes-256-gcm', key, iv);
    decipher.setAuthTag(authTag);

    return Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString('utf8');
}

export function isEncryptedEnvelope(data: unknown): data is EncryptedEnvelope {
    if (data === null || data === undefined || typeof data !== 'object') {
        return false;
    }
    const d = data as Record<string, unknown>;
    return (
        d['encrypted'] === true &&
        d['version'] === 1 &&
        d['algorithm'] === 'aes-256-gcm' &&
        typeof d['iv'] === 'string' &&
        typeof d['authTag'] === 'string' &&
        typeof d['ciphertext'] === 'string'
    );
}

export async function loadEncryptionKey(sigDir?: string): Promise<Buffer> {
    const dir = sigDir ?? DEFAULT_SIG_DIR;
    const keyPath = path.join(dir, KEY_FILE);

    let raw: string;
    try {
        raw = await fs.readFile(keyPath, 'utf8');
    } catch (e: unknown) {
        if ((e as NodeJS.ErrnoException).code === 'ENOENT') {
            throw new Error(
                'Encryption key not found at ' + keyPath + '. Run "sig init" to generate one.',
                { cause: e },
            );
        }
        throw e;
    }

    const key = Buffer.from(raw.trim(), 'base64');
    if (key.length !== 32) {
        throw new Error(`Invalid encryption key: expected 32 bytes, got ${key.length}`);
    }

    return key;
}
